gunrock/graph/dynamic_graph/slabhash_graph_parallel_iterator.cuh

```diff
@@ -0,0 +1,146 @@
+  using ValueT = typename GraphT::ValueT;
+  using SizeT = typename GraphT::SizeT;
+
+  __device__ SlabHashGraphParallelIterator(const VertexT v, GraphT* graph)
+      : v(v), graph(graph), slab_not_cached(true) {}
+
+  __device__ SizeT size() {
+    // count all the number of slabs
+    const SizeT v_buckets_offset = graph->d_buckets_offset[v];
+    const SizeT num_buckets = graph->d_hash_context[v].getNumBuckets();
+    SizeT num_slabs = 0;
+    for (int bid = 0; bid < num_buckets; ++bid) {
+      SizeT b_edges = graph->d_edges_per_bucket[bid + v_buckets_offset];
+      num_slabs += (b_edges + graph->edgesPerSlab - 1) / graph->edgesPerSlab;
+      num_slabs = b_edges == 0 ? ++num_slabs : num_slabs;
+    }
+    return num_slabs *= graph->keysPerSlab;
+  }
+
+  __device__ VertexT neighbor(const SizeT& idx) {
+    // tuples are dereferenced starting with base slabs then collision slabs
+    SizeT slabId = (idx / graph->keysPerSlab);
+    SizeT num_buckets = graph->d_hash_context[v].getNumBuckets();
+    bool is_base_slab = slabId < num_buckets;
+    uint32_t laneId = threadIdx.x & 0x1F;
+    uint32_t lane_data = EMPTY_KEY;
+    if (is_base_slab) {
+      lane_data =
+          *(graph->d_hash_context[v].getPointerFromBucket(slabId, laneId));
+    } else {
+      // iterate to find the base slab
+      slabId -= num_buckets;  // base slabs are dereferenced by now
+      SizeT prev_collision_slabs = 0;
+      SizeT num_collision_slabs = 0;
+      SizeT bid;
+      const SizeT v_buckets_offset = graph->d_buckets_offset[v];
+      for (bid = 0; bid < num_buckets; ++bid) {
+        SizeT b_edges = graph->d_edges_per_bucket[bid + v_buckets_offset];
+        num_collision_slabs +=
+            (b_edges + graph->edgesPerSlab - 1) / graph->edgesPerSlab;
+        num_collision_slabs =
+            b_edges == 0 ? ++num_collision_slabs : num_collision_slabs;
+        num_collision_slabs--;
+
+        if (slabId >= prev_collision_slabs && slabId < num_collision_slabs) {
+          slabId -= prev_collision_slabs;
+          break;  // found the base slab
+        } else {
+          prev_collision_slabs = num_collision_slabs;
+        }
+      }
+
+      // now iterate to find the slab pointer
+      uint32_t ptr = *(graph->d_hash_context[v].getPointerFromBucket(bid, 31));
+
+      for (SizeT sid = 0; sid < slabId; sid++) {
+        ptr = *(graph->d_hash_context[v].getPointerFromSlab(ptr, 31));
+      }
+      lane_data = *(graph->d_hash_context[v].getPointerFromSlab(ptr, laneId));
+    }
+
+    nodeReader<VertexT, ValueT, isMap>::readNode(lane_data, neighborId,
+                                                 neighborVal, laneId);
+    slab_not_cached = false;
+    return neighborId;
+  }
+
+  __device__ ValueT value(const SizeT& idx) {
+    if (slab_not_cached) {
+      neighbor(idx);
+    }
+    return neighborVal;
+  }
+
+ private:
+  const VertexT v;
+  GraphT* graph;
+  bool slab_not_cached;
+  VertexT neighborId;
+  ValueT neighborVal;
```

Comment:
I recommend making it obvious that private data members are indeed private. Consider adding an `_` to private member variables (see below). I don't know if we have a "coding guidelines" or "naming conventions" for Gunrock (if we do, let's follow it, if not consider this the start :)). Also, let's be consistent with our naming (e.g., don't mix `camelCase` with `not_camel_case`).

>  private:
  const VertexT v_;
  GraphT* graph_;
  bool slab_not_cached_;
  VertexT neighbor_id_;
  ValueT neighbor_val_;

Reply:
Fixed in https://github.com/gunrock/gunrock/pull/752/commits/bc3c5b5d986087863fce0b734fda79e80d38832b.